### PR TITLE
[otel-tracing] Initial opentelemetry support

### PR DESCRIPTION
**Pull request by lengrongfu**
- Closes https://github.com/distribution/distribution/issues/3451

Local Repose workflows result : 
image

### CODE REVIEW

**Comment by aaronlehmann** [comment]
I hope this can get reviewed soon. It would be really great to have this functionality in the registry.

cc @milosgajdos

**Comment by milosgajdos** [comment]
I'm aware of this. It's on my list to review indeed.

**Comment by lengrongfu** [comment]
Are the functions that need to be developed for each version discussed and announced?  I hope to have more contributions and participation.
@milosgajdos

File: registry/storage/driver/filesystem/driver.go [hunk 1]
```diff
@@ -149,9 +161,15 @@ func (d *driver) PutContent(ctx context.Context, subPath string, contents []byte
 // Reader retrieves an io.ReadCloser for the content stored at "path" with a
 // given byte offset.
 func (d *driver) Reader(ctx context.Context, path string, offset int64) (io.ReadCloser, error) {
+	span, _ := tracing.StartSpan(ctx, fmt.Sprintf("%s:%s", driverName, "Reader"),
+		trace.WithAttributes(attribute.String("path", path),
```
Comment (milosgajdos) [inline registry/storage/driver/filesystem/driver.go:165]: we should capitalise "path" or change `GetContent` so we are consistent across all recorded spans/codebase.

**Review by milosgajdos — COMMENTED** [review]
Ok, this is a fantastic effort. Thanks for the PR.

I left some comments in.

A couple of things that are missing:
* We need tests for otel `configuration` code -- there are none
* We need tests for `tracing` package -- there are none
* We need to be consistent in naming spans -- some are lowercase, some are capitalised and others are camelCase starting with lowercase character. We should be consistent across the whole codebse.

**Comment by lengrongfu** [comment]
Thanks for your comments, I will make improvements.

**Comment by lengrongfu** [comment]
> @lengrongfu seems like this is failing a lint check on `tracing/tracing_test.go` file.

Thanks,I will fix it.

File: configuration/configuration.go [hunk 2]
```diff
@@ -693,3 +714,31 @@ func Parse(rd io.Reader) (*Configuration, error) {
...
+	case ExporterTypeOTLP:
+		if otel.ServiceName == "" {
+			otel.ServiceName = ServiceName
+		}
+		if otel.TraceSamplingRatio == 0 {
+			otel.TraceSamplingRatio = DefaultSamplingRatio
+		}
+		return nil
+	default:
+		return errors.Wrapf(errors.New("invalid argument"), "unsupported exporter: %+v", otel)
```
Comment (thaJeztah) [inline configuration/configuration.go:742]: Wondering why this is using `errors.Wrapf()`; the error that's wrapped doesn't appear to have a type associated, so probably can't be used for detection here?

Comment (milosgajdos) [reply configuration/configuration.go:742]: Great catch, Seb! Yeah, we should just return `fmt.Errorf` here. No need to wrap anything since it's all plain strings.

**Comment by milosgajdos** [comment]
this is failing vendor validation @lengrongfu

File: configuration/configuration.go [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
-	"errors"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"net/http"
 	"reflect"
 	"strings"
 	"time"
+
+	"github.com/pkg/errors"
```
Comment (milosgajdos) [inline configuration/configuration.go:12]: I dont believe we need this here. We can stick with the standard library `errors` package here as we do not take anything out of this package in this file sa far as I can tell.

File: go.mod [hunk 4]
```diff
@@ -16,24 +16,33 @@ require (
 	github.com/gorilla/handlers v1.5.1
 	github.com/gorilla/mux v1.8.0
 	github.com/hashicorp/golang-lru v0.5.4
-	github.com/mitchellh/mapstructure v1.1.2
+	github.com/mitchellh/mapstructure v1.5.0
 	github.com/ncw/swift v1.0.47
 	github.com/opencontainers/go-digest v1.0.0
 	github.com/opencontainers/image-spec v1.0.2
-	github.com/prometheus/client_golang v1.12.1 // indirect; updated to latest
+	github.com/pkg/errors v0.9.1
```
Comment (milosgajdos) [inline go.mod:23]: From what I can see the only reason why we are introducing this module is we wrap errors in `tracing` pacakge and we only do it thrice, which makes me feel like we don't need to introduce this dependency and could simply use the standard library wrapping i.e. `fmt.Errorf("foo bar: %w", err)`

**Review by milosgajdos — COMMENTED** [review]
I only have a couple of comments about a newly introduced dependency but I think this otherwise looks great and once we've addressed my comments I'd be happy to push for merging this through :)

**Review by milosgajdos — APPROVED** [review]
ok, I think this LGTM. @thaJeztah, @wy65701436  can you have a look, please, in case I missed something?

File: configuration/configuration.go [hunk 5]
```diff
@@ -247,6 +247,26 @@ type Configuration struct {
...
+}
+
+// OpenTelemetryConfig provides open telemetry configuration
+type OpenTelemetryConfig struct {
+	// ServiceName is telemetry define one server name
+	ServiceName string `yaml:"serviceName,omitempty"`
+	// ExporterName must be otlp
+	ExporterName string `yaml:"exporterName"`
+	// ExporterEndpoint is otlp server addr
+	ExporterEndpoint string `yaml:"exporterEndpoint"`
```
Comment (sagikazarmark) [inline configuration/configuration.go:262]: Would it make sense to add another level of config for exporter params?

There are other configuration that can be passed to the exporter. If you ever want to expose those, you would have to prefix all of them with `Exporter` which smells IMO.

Comment (milosgajdos) [reply configuration/configuration.go:262]: I think sticking `Exporter` params into a dedicated `struct` makes sense. @lengrongfu would you mind making that change?

Comment (milosgajdos) [reply configuration/configuration.go:262]: ping @lengrongfu

**Comment by milosgajdos** [comment]
@lengrongfu this will need a rebase

**Comment by milosgajdos** [comment]
@lengrongfu seems your rebase didn't go through well -- there are some leftover chars marking conflicts

**Comment by lengrongfu** [comment]
Please look again. code conflicts have been resolved multiple times. :blush:
@milosgajdos

**Comment by milosgajdos** [comment]
This will need 👍 from @thaJeztah as he was the one requesting changes

**Comment by milosgajdos** [comment]
@lengrongfu can you rebase, please?

**Comment by milosgajdos** [comment]
ping @lengrongfu , can you rebase, please.

File: registry/storage/cache/memory/memory.go [hunk 6]
```diff
@@ -57,6 +65,10 @@ func (imbdcp *inMemoryBlobDescriptorCacheProvider) RepositoryScoped(repo string)
 }
 
 func (imbdcp *inMemoryBlobDescriptorCacheProvider) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
+	span, _ := tracing.StartSpan(ctx, fmt.Sprintf("%s:%s", TraceSpanCacheType, "Stat"),
```
Comment (corhere) [inline registry/storage/cache/memory/memory.go:68]: Concatenate string constants with the `+` operator, not `fmt.Sprintf`, so the compiler can evaluate the concatenation at build time. And do not discard the span context. It should always overwrite the ctx argument so it is always passed into called functions.
```suggestion
	span, ctx := tracing.StartSpan(ctx, TraceSpanCacheType +":Stat",
```

**Review by corhere — CHANGES_REQUESTED** [review]
I'd like to review the manual instrumentation of each package, but GitHub is chugging badly with the size of the commit. Could you please split the manual instrumentation for each package into separate commits from the commit which adds the OpenTelemetry library in?

File: tracing/tracing_test.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	"fmt"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/distribution/distribution/v3/configuration"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply tracing/tracing_test.go:14]: This does not actually bother me. I've always had the beef with `check.v1` module and was hoping to ditch it properly and figured `testify` would find its way in. We can rework this to be consistent. Thanks for bringing this up!

[the remark this answers is not in the document]
Comment (thaJeztah) [reply tracing/tracing_test.go:14]: I coincidentally was looking at that before I saw this comment https://github.com/distribution/distribution/pull/3882#issuecomment-1528989027

I agree that it's not ideal to now have different frameworks, but agree with @milosgajdos that we should replace `gopkg.in/check.v1`, and from that perspective I think it's good to not introduce more uses of that legacy framework in net-new code.

I created a ticket for this work (added "help-wanted" in case someone wants to start working on it);

- https://github.com/distribution/distribution/issues/3884

File: registry/handlers/app.go [hunk 8]
```diff
@@ -102,6 +108,10 @@ func NewApp(ctx context.Context, config *configuration.Configuration) *App {
 		isCache: config.Proxy.RemoteURL != "",
 	}
 
+	if err := app.traceHTTPMiddleware(); err != nil {
+		panic(err)
+	}
```
Comment (thaJeztah) [inline registry/handlers/app.go:113]: I assume we need to `panic()` here because `NewApp()` doesn't have an error return; perhaps we should consider changing its signature so that we can return the error instead.

**Comment by DavidSpek** [comment]
@lengrongfu it seems like something went wrong with rebasing. I don’t think you meant to close this PR.
